### PR TITLE
._glyph does not exists anymore, go for the new _getGlyphAndOffset() …

### PR DESCRIPTION
**Pull request by typemytype**
…callback

this will not be backwards compatible with older fontTools...

@jenskutilek

### CODE REVIEW

**Comment by justvanrossum** [comment]
You should not use those internal attrs/methods at all. It was previously wrong, and this PR doesn't fix it. Do not use internal stuff.

If you need access to a glyph object from the `glyf` table, you should get it from the `glyf` table.

**Review by jenskutilek — CHANGES_REQUESTED** [review]
Thank you @typemytype, I have been bitten by this change in a couple of my own modules ...

File: Lib/extractor/formats/opentype.py [hunk 1]
```diff
@@ -146,7 +146,7 @@ def extractGlyphPrograms(source, destination):
         return
     glyph_set = source.getGlyphSet()
     for name in glyph_set.keys():
-        glyph = glyph_set[name]._glyph
+        glyph, offset = glyph_set[name]._getGlyphAndOffset()
```
Comment (jenskutilek) [inline Lib/extractor/formats/opentype.py:149]: glyph = glyph_set.glyfTable[name]

**Comment by jenskutilek** [comment]
And probably you should raise the fonttools dependency in setup to `fontTools>=4.37.3`

File: Lib/extractor/formats/opentype.py (hunk 1 above)
Comment (justvanrossum) [reply Lib/extractor/formats/opentype.py:149]: No, just get the `glyf` table from the ttfont. Why even use getGlyphSet at all?

Comment (justvanrossum) [reply Lib/extractor/formats/opentype.py:149]: `glyph_set.glyfTable` should also be private.

**Comment by justvanrossum** [comment]
Just do this:
```python
def extractGlyphPrograms(source, destination):
    """
    Extract the TrueType pre-program to the font lib.
    """
    if "glyf" not in source:
        return
    glyph_table = source["glyf"]
    for name in glyph_table.keys():
        glyph = glyph_table[name]
        dest_glyph = destination[name]
        if glyph.isComposite():
            # Extract composite flags
            _extractCompositeFlags(glyph, dest_glyph)
        if not hasattr(glyph, "program"):
            continue

        hash_pen = HashPointPen(dest_glyph.width, destination)
        dest_glyph.drawPoints(hash_pen)
        lib = dest_glyph.lib[TRUETYPE_INSTRUCTIONS_KEY] = {
            "formatVersion": "1",
            "id": hash_pen.hash,
        }
        lib["assembly"] = _byteCodeToTtxAssembly(glyph.program)
```

`font.getGlyphSet()` is an abstraction to extract outline data, _not_ a way to get at raw `glyf glyphs.

**Comment by jenskutilek** [comment]
I was probably under the impression that the glyphset was the preferred way of getting at the glyphs because `getGlyphSet()` is so high up in the module ...

**Comment by jenskutilek** [comment]
> And probably you should raise the fonttools dependency in setup to `fontTools>=4.37.3`

With @justvanrossum’s suggestion, the dependency can of course stay as it was.

**Comment by typemytype** [comment]
thanks @justvanrossum

**Comment by justvanrossum** [comment]
> I was probably under the impression that the glyphset was the preferred way of getting at the glyphs because getGlyphSet() is so high up in the module ...

But using attrs starting with underscores should trigger all alarm bells...

`font.getGlyphSet()` is high up because it implements drawable glyphs for both `glyf` and `CFF ` (*), and abstracts away their differences where possible. If you need to access the `glyf` table, you should access the `glyf` table, and not try to sneak your way through a higher level abstraction.

*) And since recently (related to the refactoring that triggered this PR) it supports `glyf` + `gvar` and `CFF2`, via the `location` argument of `font.getGlyphSet()`.

**Pull request merged**